File: android/src/main/java/co/broadcastapp/muckabout/BffApiClient.java
package co.broadcastapp.muckabout;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

/**
 * Native HTTP client for fetching media browse data from the WNYC BFF API.
 * Used by RemoteStreamerService to populate the Android Auto media browse tree.
 */
public class BffApiClient {
    private static final String TAG = "BffApiClient";
    private static final int TIMEOUT_MS = 10000;

    private String baseUrl;

    public BffApiClient(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    public void setBaseUrl(String baseUrl) {
        this.baseUrl = baseUrl;
    }

    // --- Data classes ---

    public static class LiveStream {
        public String slug;
        public String stationName;
        public String currentShowTitle;
        public String hlsUrl;
        public String imageUrl;

        public LiveStream(String slug, String stationName, String currentShowTitle, String hlsUrl, String imageUrl) {
            this.slug = slug;
            this.stationName = stationName;
            this.currentShowTitle = currentShowTitle;
            this.hlsUrl = hlsUrl;
            this.imageUrl = imageUrl;
        }
    }

    public static class NewsItem {
        public String id;
        public String title;
        public String showTitle;
        public String audioUrl;
        public String imageUrl;
        public int durationSeconds;

        public NewsItem(String id, String title, String showTitle, String audioUrl, String imageUrl, int durationSeconds) {
            this.id = id;
            this.title = title;
            this.showTitle = showTitle;
            this.audioUrl = audioUrl;
            this.imageUrl = imageUrl;
            this.durationSeconds = durationSeconds;
        }
    }

    public static class StoryItem {
        public String id;
        public String title;
        public String showTitle;
        public String audioUrl;
        public String imageUrl;
        public int durationSeconds;

        public StoryItem(String id, String title, String showTitle, String audioUrl, String imageUrl, int durationSeconds) {
            this.id = id;
            this.title = title;
            this.showTitle = showTitle;
            this.audioUrl = audioUrl;
            this.imageUrl = imageUrl;
            this.durationSeconds = durationSeconds;
        }
    }

    public static class Show {
        public String slug;
        public String title;
        public String imageUrl;

        public Show(String slug, String title, String imageUrl) {
            this.slug = slug;
            this.title = title;
            this.imageUrl = imageUrl;
        }
    }

    public static class Episode {
        public String id;
        public String title;
        public String showTitle;
        public String audioUrl;
        public String imageUrl;
        public int durationSeconds;

        public Episode(String id, String title, String showTitle, String audioUrl, String imageUrl, int durationSeconds) {
            this.id = id;
            this.title = title;
            this.showTitle = showTitle;
            this.audioUrl = audioUrl;
            this.imageUrl = imageUrl;
            this.durationSeconds = durationSeconds;
        }
    }

    // --- API methods ---

    /**
     * Fetch live streams from /api/streams
     */
    public List<LiveStream> fetchLiveStreams() {
        List<LiveStream> streams = new ArrayList<>();
        try {
            String json = httpGet(baseUrl + "/api/streams");
            if (json == null) return streams;

            JSONArray arr = new JSONArray(json);
            for (int i = 0; i < arr.length(); i++) {
                JSONObject obj = arr.getJSONObject(i);
                String slug = obj.optString("slug", "");
                String stationName = obj.optString("station", "");
                String currentShow = obj.optString("title", obj.optString("showTitle", ""));

                // Image is a direct URL string
                String imageUrl = obj.optString("image", "");
                if (imageUrl.isEmpty()) {
                    // Fallback to stationImage.url
                    JSONObject stationImage = obj.optJSONObject("stationImage");
                    if (stationImage != null) {
                        String template = stationImage.optString("template", "");
                        if (!template.isEmpty()) {
                            imageUrl = template.replace("%s/%s/%s/%s", "512/512/c/80");
                        } else {
                            imageUrl = stationImage.optString("url", "");
                        }
                    }
                }

                // HLS URL
                String hlsUrl = obj.optString("hls", "");
                if (hlsUrl.isEmpty()) {
                    hlsUrl = obj.optString("audio", obj.optString("file", ""));
                }

                if (!hlsUrl.isEmpty()) {
                    streams.add(new LiveStream(slug, stationName, currentShow, hlsUrl, imageUrl));
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing live streams", e);
        }
        return streams;
    }

    /**
     * Fetch latest news from /api/homepagelatestnewsupdates
     */
    public List<NewsItem> fetchLatestNews() {
        List<NewsItem> news = new ArrayList<>();
        try {
            String json = httpGet(baseUrl + "/api/homepagelatestnewsupdates");
            if (json == null) return news;

            JSONObject obj = new JSONObject(json);

            // Local newscast (NYC Headlines)
            JSONObject local = obj.optJSONObject("local_newscast");
            if (local == null) local = obj.optJSONObject("localNewscast");
            if (local != null) {
                news.add(parseNewsItem(local, "local_newscast"));
            }

            // National newscast (NPR News Now)
            JSONObject national = obj.optJSONObject("national_newscast");
            if (national == null) national = obj.optJSONObject("nationalNewscast");
            if (national != null) {
                news.add(parseNewsItem(national, "national_newscast"));
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing latest news", e);
        }
        return news;
    }

    /**
     * Fetch top stories from /api/curated_lists/4 (demo) or /api/curated_lists/87 (prod)
     * Returns up to 4 items that have audio
     */
    public List<StoryItem> fetchTopStories() {
        List<StoryItem> stories = new ArrayList<>();
        try {
            // Using list ID 4 for demo, 87 for prod
            String json = httpGet(baseUrl + "/api/curated_lists/87");
            if (json == null) return stories;

            JSONObject obj = new JSONObject(json);
            JSONArray listItems = obj.optJSONArray("listItems");
            if (listItems == null) listItems = obj.optJSONArray("list_items");
            if (listItems == null) return stories;

            for (int i = 0; i < listItems.length() && stories.size() < 4; i++) {
                JSONObject item = listItems.getJSONObject(i);

                // Only include items that explicitly have audio
                if (!item.optBoolean("hasAudio", false)) continue;
                String audioUrl = item.optString("audio", "");
                if (audioUrl.isEmpty()) continue;

                String id = item.optString("id", "story_" + i);
                String title = item.optString("title", "");
                String showTitle = item.optString("showTitle", item.optString("show_title", ""));
                int duration = item.optInt("estimatedDuration", item.optInt("estimated_duration", 0));

                String imageUrl = resolveImageUrl(item, true, "image");

                stories.add(new StoryItem(id, title, showTitle, audioUrl, imageUrl, duration));
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing top stories", e);
        }
        return stories;
    }

    /**
     * Fetch all shows from /api/v2/discover/shows
     */
    public List<Show> fetchAllShows() {
        List<Show> shows = new ArrayList<>();
        try {
            String json = httpGet(baseUrl + "/api/v3/shows");
            if (json == null) return shows;

            JSONObject obj = new JSONObject(json);
            // just interested in the featured shows for now, which is what Android Auto will show
            JSONArray allShows = obj.optJSONArray("all");
            if (allShows == null) return shows;

            for (int i = 0; i < allShows.length(); i++) {
                JSONObject showObj = allShows.getJSONObject(i);
                String slug = showObj.optString("slug", showObj.optString("id", ""));
                // Primary shape is top-level fields (title, image, showArt), with legacy attrs fallback.
                String title = showObj.optString("title", "");
                String imageUrl = resolveImageUrl(showObj, false, "image", "showArt", "logoImage", "logo_image");

                if (title.isEmpty() || imageUrl.isEmpty()) {
                    JSONObject attrs = showObj.optJSONObject("attributes");
                    if (attrs != null) {
                        if (title.isEmpty()) {
                            title = attrs.optString("title", "");
                        }
                        if (imageUrl.isEmpty()) {
                            imageUrl = resolveImageUrl(attrs, false, "image", "image-main", "imageMain");
                        }
                    }
                }

                if (!title.isEmpty()) {
                    shows.add(new Show(slug, title, imageUrl));
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing shows", e);
        }
        return shows;
    }

    /**
     * Fetch episodes for a show. First resolves the podcastId from the show page,
     * then fetches episodes using that podcastId.
     */
    public List<Episode> fetchEpisodes(String showSlug) {
        List<Episode> episodes = new ArrayList<>();
        try {
            // Step 1: Fetch show page to get podcastId from linkedDataSource
            String showJson = httpGet(baseUrl + "/api/pages/wagtail/" + showSlug + "?showOnly=true");
            if (showJson == null) return episodes;

            JSONObject showObj = new JSONObject(showJson);
            JSONArray linkedDataSource = showObj.optJSONArray("linkedDataSource");
            if (linkedDataSource == null || linkedDataSource.length() == 0) return episodes;

            JSONObject firstSource = linkedDataSource.getJSONObject(0);
            JSONObject value = firstSource.optJSONObject("value");
            if (value == null) return episodes;

            String podcastId = value.optString("id", "");
            if (podcastId.isEmpty()) return episodes;

            // Step 2: Fetch episodes using the podcastId
            String json = httpGet(baseUrl + "/api/v3/show/" + podcastId + "/episodes?limit=20");
            if (json == null) return episodes;

            JSONObject obj = new JSONObject(json);
            JSONArray data = obj.optJSONArray("data");
            if (data == null) {
                // Might be a direct array
                try {
                    data = new JSONArray(json);
                } catch (JSONException ignored) {
                    return episodes;
                }
            }

            for (int i = 0; i < data.length(); i++) {
                JSONObject ep = data.getJSONObject(i);
                JSONObject attrs = ep.optJSONObject("attributes");
                if (attrs == null) attrs = ep;

                String id = ep.optString("id", "ep_" + i);
                String title = attrs.optString("title", ep.optString("title", ""));
                String showTitle = attrs.optString("show_title", attrs.optString("showTitle", ep.optString("showTitle", "")));
                String audioUrl = attrs.optString("audio", attrs.optString("file", ep.optString("audio", ep.optString("file", ""))));
                String imageUrl = resolveImageUrl(attrs, false, "image", "showArt", "logoImage", "logo_image");
                if (imageUrl.isEmpty()) {
                    imageUrl = resolveImageUrl(ep, false, "image", "showArt", "logoImage", "logo_image");
                }
                int duration = attrs.optInt("duration", attrs.optInt("estimated_duration", 0));

                if (!audioUrl.isEmpty()) {
                    episodes.add(new Episode(id, title, showTitle, audioUrl, imageUrl, duration));
                }
            }
        } catch (JSONException e) {
            Log.e(TAG, "Error parsing episodes for " + showSlug, e);
        }
        return episodes;
    }

    // --- Helpers ---

    private NewsItem parseNewsItem(JSONObject obj, String fallbackId) throws JSONException {
        String id = obj.optString("id", fallbackId);
        String title = obj.optString("card_title", obj.optString("cardTitle", obj.optString("title", "")));
        String showTitle = obj.optString("show_title", obj.optString("showTitle", ""));
        String audioUrl = obj.optString("file", obj.optString("audio", ""));
        String imageUrl = "";
        JSONObject headers = obj.optJSONObject("headers");
        if (headers != null) {
            JSONObject brand = headers.optJSONObject("brand");
            if (brand != null) {
                JSONObject logoImage = brand.optJSONObject("logoImage");
                if (logoImage != null) {
                    imageUrl = normalizeImageUrl(logoImage.optString("url", ""));
                }
            }
        }
        int duration = obj.optInt("duration", obj.optInt("estimated_duration", 0));

        return new NewsItem(id, title, showTitle, audioUrl, imageUrl, duration);
    }

    private String resolveImageUrl(JSONObject container, boolean includeBrandFallback, String... imageKeys) {
        return resolveImageUrl(container, includeBrandFallback, false, imageKeys);
    }

    private String resolveImageUrl(JSONObject container, boolean includeBrandFallback, boolean forceFallback, String... imageKeys) {
        if (container == null) return "";

        if (!forceFallback) {
            for (String imageKey : imageKeys) {
                String imageUrl = resolveImageCandidate(container.opt(imageKey));
                if (!imageUrl.isEmpty()) {
                    return imageUrl;
                }
            }
        }

        if (includeBrandFallback) {
            return resolveBrandLogoUrl(container);
        }

        return "";
    }

    private String resolveImageCandidate(Object imageObj) {
        if (imageObj instanceof JSONObject) {
            JSONObject image = (JSONObject) imageObj;
            String imageUrl = image.optString("file", image.optString("url", ""));
            if (imageUrl.isEmpty()) {
                String template = image.optString("template", "");
                if (!template.isEmpty()) {
                    imageUrl = template.replace("%s/%s/%s/%s", "200/200/c/80");
                }
            }
            return normalizeImageUrl(imageUrl);
        }

        if (imageObj instanceof String) {
            return normalizeImageUrl((String) imageObj);
        }

        return "";
    }

    private String resolveBrandLogoUrl(JSONObject container) {
        JSONObject headers = container.optJSONObject("headers");
        if (headers == null) return "";

        JSONObject brand = headers.optJSONObject("brand");
        if (brand == null) return "";

        JSONObject logoImage = brand.optJSONObject("logoImage");
        if (logoImage == null) logoImage = brand.optJSONObject("logo_image");
        if (logoImage == null) return "";

        return normalizeImageUrl(logoImage.optString("url", ""));
    }

    private String normalizeImageUrl(String imageUrl) {
        if (imageUrl == null || imageUrl.isEmpty()) return "";

        if (imageUrl.contains("npr.brightspotcdn.com")) {
            return imageUrl.replace("{width}", "512")
                    .replace("{quality}", "80")
                    .replace("{format}", "jpg");
        }

        return imageUrl;
    }

    private String httpGet(String urlStr) {
        HttpURLConnection connection = null;
        try {
            URL url = new URL(urlStr);
            connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(TIMEOUT_MS);
            connection.setReadTimeout(TIMEOUT_MS);
            connection.setRequestProperty("Accept", "application/json");
            connection.setInstanceFollowRedirects(true);

            int code = connection.getResponseCode();
            if (code != 200) {
                Log.w(TAG, "HTTP " + code + " from " + urlStr);
                return null;
            }

            BufferedReader reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder sb = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                sb.append(line);
            }
            reader.close();
            return sb.toString();
        } catch (IOException e) {
            Log.e(TAG, "HTTP request failed: " + urlStr, e);
            return null;
        } finally {
            if (connection != null) connection.disconnect();
        }
    }
}

File: ios/Sources/RemoteStreamerPlugin/BffApiClient.swift
import Foundation

/// Native HTTP client for fetching media browse data from the WNYC BFF API.
/// Used by CarPlay to populate the browse tree — mirrors the Android BffApiClient.
@available(iOS 14.0, *)
class BffApiClient {
    static let shared = BffApiClient()

    private var baseUrl = "https://wnyc.org"
    private var aviaryBaseUrl = "https://cms.nypr.digital/api/v2"
    private let session: URLSession
    private let timeoutInterval: TimeInterval = 10

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeoutInterval
        config.timeoutIntervalForResource = timeoutInterval
        session = URLSession(configuration: config)
    }

    func setBaseUrl(_ url: String) {
        baseUrl = url
    }

    // MARK: - Data Models

    struct LiveStream {
        let slug: String
        let stationName: String
        let currentShowTitle: String
        let hlsUrl: String
        let imageUrl: String
    }

    struct NewsItem {
        let id: String
        let title: String
        let showTitle: String
        let audioUrl: String
        let imageUrl: String
        let durationSeconds: Int
    }

    struct StoryItem {
        let id: String
        let title: String
        let showTitle: String
        let audioUrl: String
        let imageUrl: String
        let durationSeconds: Int
    }

    struct Show {
        let slug: String
        let title: String
        let imageUrl: String
    }

    struct Episode {
        let id: String
        let title: String
        let showTitle: String
        let audioUrl: String
        let imageUrl: String
        let durationSeconds: Int
    }

    // MARK: - API Methods

    /// Fetch live streams from /api/streams
    func fetchLiveStreams() -> [LiveStream] {
        guard let json = httpGetSync(baseUrl + "/api/streams"),
              let arr = try? JSONSerialization.jsonObject(with: json) as? [[String: Any]] else {
            return []
        }

        var streams: [LiveStream] = []
        for obj in arr {
            let slug = obj["slug"] as? String ?? ""
            let stationName = obj["station"] as? String ?? ""
            let currentShow = obj["title"] as? String ?? obj["showTitle"] as? String ?? ""

            var imageUrl = obj["image"] as? String ?? ""
            if imageUrl.isEmpty {
                if let stationImage = obj["stationImage"] as? [String: Any] {
                    let template = stationImage["template"] as? String ?? ""
                    if !template.isEmpty {
                        imageUrl = template.replacingOccurrences(of: "%s/%s/%s/%s", with: "512/512/c/80")
                    } else {
                        imageUrl = stationImage["url"] as? String ?? ""
                    }
                }
            }

            var hlsUrl = obj["hls"] as? String ?? ""
            if hlsUrl.isEmpty {
                hlsUrl = obj["audio"] as? String ?? obj["file"] as? String ?? ""
            }

            if !hlsUrl.isEmpty {
                streams.append(LiveStream(slug: slug, stationName: stationName, currentShowTitle: currentShow, hlsUrl: hlsUrl, imageUrl: imageUrl))
            }
        }
        return streams
    }

    /// Fetch latest news from /api/homepagelatestnewsupdates
    func fetchLatestNews() -> [NewsItem] {
        guard let json = httpGetSync(baseUrl + "/api/homepagelatestnewsupdates"),
              let obj = try? JSONSerialization.jsonObject(with: json) as? [String: Any] else {
            return []
        }

        var news: [NewsItem] = []

        // Local newscast (NYC Headlines)
        let local = obj["local_newscast"] as? [String: Any] ?? obj["localNewscast"] as? [String: Any]
        if let local = local, let item = parseNewsItem(local, fallbackId: "local_newscast") {
            news.append(item)
        }

        // National newscast (NPR News Now)
        let national = obj["national_newscast"] as? [String: Any] ?? obj["nationalNewscast"] as? [String: Any]
        if let national = national, let item = parseNewsItem(national, fallbackId: "national_newscast") {
            news.append(item)
        }

        return news
    }

    /// Fetch top stories from /api/curated_lists/87
    func fetchTopStories() -> [StoryItem] {
        guard let json = httpGetSync(baseUrl + "/api/curated_lists/87"),
              let obj = try? JSONSerialization.jsonObject(with: json) as? [String: Any] else {
            return []
        }

        let listItems = obj["listItems"] as? [[String: Any]] ?? obj["list_items"] as? [[String: Any]] ?? []

        var stories: [StoryItem] = []
        for item in listItems {
            guard stories.count < 4 else { break }

            let hasAudio = item["hasAudio"] as? Bool ?? false
            guard hasAudio else { continue }

            let audioUrl = item["audio"] as? String ?? ""
            guard !audioUrl.isEmpty else { continue }

            let id = item["id"] as? String ?? "story_\(stories.count)"
            let title = item["title"] as? String ?? ""
            let showTitle = item["showTitle"] as? String ?? item["show_title"] as? String ?? ""
            let duration = item["estimatedDuration"] as? Int ?? item["estimated_duration"] as? Int ?? 0
            let imageUrl = resolveImageUrl(from: item, includeBrandFallback: true, keys: "image")

            stories.append(StoryItem(id: id, title: title, showTitle: showTitle, audioUrl: audioUrl, imageUrl: imageUrl, durationSeconds: duration))
        }
        return stories
    }

    /// Fetch all shows from /api/v3/shows
    func fetchAllShows() -> [Show] {
        guard let json = httpGetSync(baseUrl + "/api/v3/shows"),
              let obj = try? JSONSerialization.jsonObject(with: json) as? [String: Any] else {
            return []
        }

        guard let allShows = obj["all"] as? [[String: Any]] else { return [] }

        var shows: [Show] = []
        for showObj in allShows {
            let slug = showObj["slug"] as? String ?? showObj["id"] as? String ?? ""
            var title = showObj["title"] as? String ?? ""
            var imageUrl = resolveImageUrl(from: showObj, includeBrandFallback: false, keys: "image", "showArt", "logoImage", "logo_image")

            if title.isEmpty || imageUrl.isEmpty {
                if let attrs = showObj["attributes"] as? [String: Any] {
                    if title.isEmpty {
                        title = attrs["title"] as? String ?? ""
                    }
                    if imageUrl.isEmpty {
                        imageUrl = resolveImageUrl(from: attrs, includeBrandFallback: false, keys: "image", "image-main", "imageMain")
                    }
                }
            }

            if !title.isEmpty {
                shows.append(Show(slug: slug, title: title, imageUrl: imageUrl))
            }
        }
        return shows
    }

    /// Fetch episodes for a show
    func fetchEpisodes(showSlug: String) -> [Episode] {
        // Step 1: Fetch show page to get podcastId from linkedDataSource
        guard let showJson = httpGetSync(baseUrl + "/api/pages/wagtail/" + showSlug + "?showOnly=true"),
              let showObj = try? JSONSerialization.jsonObject(with: showJson) as? [String: Any] else {
            return []
        }

        guard let linkedDataSource = showObj["linkedDataSource"] as? [[String: Any]],
              let firstSource = linkedDataSource.first,
              let value = firstSource["value"] as? [String: Any],
              let podcastId = value["id"] as? String, !podcastId.isEmpty else {
            return []
        }

        // Step 2: Fetch episodes using the podcastId
        guard let json = httpGetSync(baseUrl + "/api/v3/show/" + podcastId + "/episodes?limit=20"),
              let obj = try? JSONSerialization.jsonObject(with: json) as? [String: Any] else {
            return []
        }

        let data = obj["data"] as? [[String: Any]] ?? []

        var episodes: [Episode] = []
        for ep in data {
            let attrs = ep["attributes"] as? [String: Any] ?? ep
            let id = ep["id"] as? String ?? "ep_\(episodes.count)"
            let title = attrs["title"] as? String ?? ep["title"] as? String ?? ""
            let showTitle = attrs["show_title"] as? String ?? attrs["showTitle"] as? String ?? ep["showTitle"] as? String ?? ""
            let audioUrl = attrs["audio"] as? String ?? attrs["file"] as? String ?? ep["audio"] as? String ?? ep["file"] as? String ?? ""
            var imageUrl = resolveImageUrl(from: attrs, includeBrandFallback: false, keys: "image", "showArt", "logoImage", "logo_image")
            if imageUrl.isEmpty {
                imageUrl = resolveImageUrl(from: ep, includeBrandFallback: false, keys: "image", "showArt", "logoImage", "logo_image")
            }
            let duration = attrs["duration"] as? Int ?? attrs["estimated_duration"] as? Int ?? 0

            if !audioUrl.isEmpty {
                episodes.append(Episode(id: id, title: title, showTitle: showTitle, audioUrl: audioUrl, imageUrl: imageUrl, durationSeconds: duration))
            }
        }
        return episodes
    }

    // MARK: - Helpers

    private func parseNewsItem(_ obj: [String: Any], fallbackId: String) -> NewsItem? {
        let id = obj["id"] as? String ?? fallbackId
        let title = obj["card_title"] as? String ?? obj["cardTitle"] as? String ?? obj["title"] as? String ?? ""
        let showTitle = obj["show_title"] as? String ?? obj["showTitle"] as? String ?? ""
        let audioUrl = obj["file"] as? String ?? obj["audio"] as? String ?? ""
        var imageUrl = ""
        if let headers = obj["headers"] as? [String: Any],
           let brand = headers["brand"] as? [String: Any],
           let logoImage = brand["logoImage"] as? [String: Any] {
            imageUrl = normalizeImageUrl(logoImage["url"] as? String ?? "")
        }
        let duration = obj["duration"] as? Int ?? obj["estimated_duration"] as? Int ?? 0

        guard !audioUrl.isEmpty else { return nil }
        return NewsItem(id: id, title: title, showTitle: showTitle, audioUrl: audioUrl, imageUrl: imageUrl, durationSeconds: duration)
    }

    private func resolveImageUrl(from container: [String: Any], includeBrandFallback: Bool, keys: String...) -> String {
        for key in keys {
            let candidate = resolveImageCandidate(container[key])
            if !candidate.isEmpty { return candidate }
        }
        if includeBrandFallback {
            return resolveBrandLogoUrl(from: container)
        }
        return ""
    }

    private func resolveImageCandidate(_ imageObj: Any?) -> String {
        if let image = imageObj as? [String: Any] {
            var imageUrl = image["file"] as? String ?? image["url"] as? String ?? ""
            if imageUrl.isEmpty {
                let template = image["template"] as? String ?? ""
                if !template.isEmpty {
                    imageUrl = template.replacingOccurrences(of: "%s/%s/%s/%s", with: "200/200/c/80")
                }
            }
            return normalizeImageUrl(imageUrl)
        }
        if let imageStr = imageObj as? String {
            return normalizeImageUrl(imageStr)
        }
        return ""
    }

    private func resolveBrandLogoUrl(from container: [String: Any]) -> String {
        guard let headers = container["headers"] as? [String: Any],
              let brand = headers["brand"] as? [String: Any] else { return "" }
        let logoImage = brand["logoImage"] as? [String: Any] ?? brand["logo_image"] as? [String: Any]
        guard let logo = logoImage else { return "" }
        return normalizeImageUrl(logo["url"] as? String ?? "")
    }

    private func normalizeImageUrl(_ imageUrl: String) -> String {
        guard !imageUrl.isEmpty else { return "" }
        if imageUrl.contains("npr.brightspotcdn.com") {
            return imageUrl
                .replacingOccurrences(of: "{width}", with: "512")
                .replacingOccurrences(of: "{quality}", with: "80")
                .replacingOccurrences(of: "{format}", with: "jpg")
        }
        return imageUrl
    }

    // MARK: - HTTP

    private func httpGetSync(_ urlStr: String) -> Data? {
        guard let url = URL(string: urlStr) else { return nil }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.timeoutInterval = timeoutInterval

        var result: Data?
        let semaphore = DispatchSemaphore(value: 0)

        let task = session.dataTask(with: request) { data, response, error in
            if let error = error {
                print("BffApiClient: HTTP request failed: \(urlStr) - \(error.localizedDescription)")
                semaphore.signal()
                return
            }
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                let code = (response as? HTTPURLResponse)?.statusCode ?? -1
                print("BffApiClient: HTTP \(code) from \(urlStr)")
                semaphore.signal()
                return
            }
            result = data
            semaphore.signal()
        }
        task.resume()
        semaphore.wait()
        return result
    }
}
